### ISSUE TITLE
Terraform 0.12 support issue

### ISSUE THREAD

**Issue by ax-atlantis** [terraform-aws-modules/terraform-aws-atlantis#56]
Attempting to deploy using Terraform 0.12 and running into several alarms.

Terraform implementation:
```
module "atlantis" {
  source = "terraform-aws-modules/atlantis/aws"

  name = "atlantis"

  # VPC
  vpc_id              = var.vpc_id
  private_subnet_ids  = var.private_subnet_ids
  public_subnet_ids   = var.public_subnet_ids

  # DNS (without trailing dot)
  route53_zone_name = var.zone_name

  # ACM (SSL certificate) - Specify ARN of an existing certificate or new one will be created and validated using Route53 DNS
  acm_certificate_domain_name = var.domain_name
  certificate_arn = var.acm_certificate_arn

  # Atlantis
  atlantis_github_user       = var.github_user
  atlantis_github_user_token_ssm_parameter_name = "/atlantis/user_token"
  atlantis_repo_whitelist    = ["github.com/terraform-aws-modules/*"]
  allow_repo_config = "true"
}
```

```
terraform apply

Error: Missing resource instance key

  on .terraform/modules/atlantis.acm/terraform-aws-modules-terraform-aws-acm-4c42634/main.tf line 29, in resource "aws_route53_record" "validation":
  29:   name    = "${aws_acm_certificate.this.domain_validation_options.0.resource_record_name}"

Because aws_acm_certificate.this has "count" set, its attributes must be
accessed on specific instances.

For example, to correlate with indices of a referring resource, use:
    aws_acm_certificate.this[count.index]


Error: Missing resource instance key

  on .terraform/modules/atlantis.acm/terraform-aws-modules-terraform-aws-acm-4c42634/main.tf line 30, in resource "aws_route53_record" "validation":
  30:   type    = "${aws_acm_certificate.this.domain_validation_options.0.resource_record_type}"

Because aws_acm_certificate.this has "count" set, its attributes must be
accessed on specific instances.

For example, to correlate with indices of a referring resource, use:
    aws_acm_certificate.this[count.index]


Error: Missing resource instance key

  on .terraform/modules/atlantis.acm/terraform-aws-modules-terraform-aws-acm-4c42634/main.tf line 34, in resource "aws_route53_record" "validation":
  34:     "${aws_acm_certificate.this.domain_validation_options.0.resource_record_value}",

Because aws_acm_certificate.this has "count" set, its attributes must be
accessed on specific instances.

For example, to correlate with indices of a referring resource, use:
    aws_acm_certificate.this[count.index]


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_http_sg/terraform-aws-modules-terraform-aws-security-group-9c0d748/main.tf line 119, in resource "aws_security_group_rule" "ingress_with_cidr_blocks":
 119:   cidr_blocks     = ["${split(",", lookup(var.ingress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.ingress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_http_sg/terraform-aws-modules-terraform-aws-security-group-9c0d748/main.tf line 135, in resource "aws_security_group_rule" "computed_ingress_with_cidr_blocks":
 135:   cidr_blocks     = ["${split(",", lookup(var.computed_ingress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.ingress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_http_sg/terraform-aws-modules-terraform-aws-security-group-9c0d748/main.tf line 151, in resource "aws_security_group_rule" "ingress_with_ipv6_cidr_blocks":
 151:   ipv6_cidr_blocks = ["${split(",", lookup(var.ingress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.ingress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_http_sg/terraform-aws-modules-terraform-aws-security-group-9c0d748/main.tf line 167, in resource "aws_security_group_rule" "computed_ingress_with_ipv6_cidr_blocks":
 167:   ipv6_cidr_blocks = ["${split(",", lookup(var.computed_ingress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.ingress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_http_sg/terraform-aws-modules-terraform-aws-security-group-9c0d748/main.tf line 295, in resource "aws_security_group_rule" "egress_with_cidr_blocks":
 295:   cidr_blocks     = ["${split(",", lookup(var.egress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.egress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_http_sg/terraform-aws-modules-terraform-aws-security-group-9c0d748/main.tf line 311, in resource "aws_security_group_rule" "computed_egress_with_cidr_blocks":
 311:   cidr_blocks     = ["${split(",", lookup(var.computed_egress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.egress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_http_sg/terraform-aws-modules-terraform-aws-security-group-9c0d748/main.tf line 327, in resource "aws_security_group_rule" "egress_with_ipv6_cidr_blocks":
 327:   ipv6_cidr_blocks = ["${split(",", lookup(var.egress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.egress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_http_sg/terraform-aws-modules-terraform-aws-security-group-9c0d748/main.tf line 343, in resource "aws_security_group_rule" "computed_egress_with_ipv6_cidr_blocks":
 343:   ipv6_cidr_blocks = ["${split(",", lookup(var.computed_egress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.egress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_https_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 119, in resource "aws_security_group_rule" "ingress_with_cidr_blocks":
 119:   cidr_blocks     = ["${split(",", lookup(var.ingress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.ingress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_https_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 135, in resource "aws_security_group_rule" "computed_ingress_with_cidr_blocks":
 135:   cidr_blocks     = ["${split(",", lookup(var.computed_ingress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.ingress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_https_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 151, in resource "aws_security_group_rule" "ingress_with_ipv6_cidr_blocks":
 151:   ipv6_cidr_blocks = ["${split(",", lookup(var.ingress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.ingress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_https_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 167, in resource "aws_security_group_rule" "computed_ingress_with_ipv6_cidr_blocks":
 167:   ipv6_cidr_blocks = ["${split(",", lookup(var.computed_ingress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.ingress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_https_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 295, in resource "aws_security_group_rule" "egress_with_cidr_blocks":
 295:   cidr_blocks     = ["${split(",", lookup(var.egress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.egress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_https_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 311, in resource "aws_security_group_rule" "computed_egress_with_cidr_blocks":
 311:   cidr_blocks     = ["${split(",", lookup(var.computed_egress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.egress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_https_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 327, in resource "aws_security_group_rule" "egress_with_ipv6_cidr_blocks":
 327:   ipv6_cidr_blocks = ["${split(",", lookup(var.egress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.egress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.alb_https_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 343, in resource "aws_security_group_rule" "computed_egress_with_ipv6_cidr_blocks":
 343:   ipv6_cidr_blocks = ["${split(",", lookup(var.computed_egress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.egress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.atlantis_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 119, in resource "aws_security_group_rule" "ingress_with_cidr_blocks":
 119:   cidr_blocks     = ["${split(",", lookup(var.ingress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.ingress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.atlantis_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 135, in resource "aws_security_group_rule" "computed_ingress_with_cidr_blocks":
 135:   cidr_blocks     = ["${split(",", lookup(var.computed_ingress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.ingress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.atlantis_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 151, in resource "aws_security_group_rule" "ingress_with_ipv6_cidr_blocks":
 151:   ipv6_cidr_blocks = ["${split(",", lookup(var.ingress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.ingress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.atlantis_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 167, in resource "aws_security_group_rule" "computed_ingress_with_ipv6_cidr_blocks":
 167:   ipv6_cidr_blocks = ["${split(",", lookup(var.computed_ingress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.ingress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.atlantis_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 295, in resource "aws_security_group_rule" "egress_with_cidr_blocks":
 295:   cidr_blocks     = ["${split(",", lookup(var.egress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.egress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.atlantis_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 311, in resource "aws_security_group_rule" "computed_egress_with_cidr_blocks":
 311:   cidr_blocks     = ["${split(",", lookup(var.computed_egress_with_cidr_blocks[count.index], "cidr_blocks", join(",", var.egress_cidr_blocks)))}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.atlantis_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 327, in resource "aws_security_group_rule" "egress_with_ipv6_cidr_blocks":
 327:   ipv6_cidr_blocks = ["${split(",", lookup(var.egress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.egress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.atlantis_sg/terraform-aws-modules-terraform-aws-security-group-0fd7c5a/main.tf line 343, in resource "aws_security_group_rule" "computed_egress_with_ipv6_cidr_blocks":
 343:   ipv6_cidr_blocks = ["${split(",", lookup(var.computed_egress_with_ipv6_cidr_blocks[count.index], "ipv6_cidr_blocks", join(",", var.egress_ipv6_cidr_blocks)))}"]

Inappropriate value for attribute "ipv6_cidr_blocks": element 0: string
required.


Error: Missing resource instance key

  on .terraform/modules/atlantis.vpc/terraform-aws-modules-terraform-aws-vpc-d59af8f/main.tf line 31, in resource "aws_vpc_ipv4_cidr_block_association" "this":
  31:   vpc_id = "${aws_vpc.this.id}"

Because aws_vpc.this has "count" set, its attributes must be accessed on
specific instances.

For example, to correlate with indices of a referring resource, use:
    aws_vpc.this[count.index]


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.vpc/terraform-aws-modules-terraform-aws-vpc-d59af8f/main.tf line 43, in resource "aws_vpc_dhcp_options" "this":
  43:   domain_name_servers  = ["${var.dhcp_options_domain_name_servers}"]

Inappropriate value for attribute "domain_name_servers": element 0: string
required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.vpc/terraform-aws-modules-terraform-aws-vpc-d59af8f/main.tf line 44, in resource "aws_vpc_dhcp_options" "this":
  44:   ntp_servers          = ["${var.dhcp_options_ntp_servers}"]

Inappropriate value for attribute "ntp_servers": element 0: string required.


Error: Incorrect attribute value type

  on .terraform/modules/atlantis.vpc/terraform-aws-modules-terraform-aws-vpc-d59af8f/main.tf line 45, in resource "aws_vpc_dhcp_options" "this":
  45:   netbios_name_servers = ["${var.dhcp_options_netbios_name_servers}"]

Inappropriate value for attribute "netbios_name_servers": element 0: string
required.


Error: Unexpected resource instance key

  on .terraform/modules/atlantis/terraform-aws-modules-terraform-aws-atlantis-3eae5e6/main.tf line 362, in data "aws_iam_policy_document" "ecs_task_access_secrets_with_kms":
 362:   source_json = "${data.aws_iam_policy_document.ecs_task_access_secrets.0.json}"

Because data.aws_iam_policy_document.ecs_task_access_secrets does not have
"count" or "for_each" set, references to it must not include an index key.
Remove the bracketed index to refer to the single instance of this resource.


Error: Invalid value for module argument

  on .terraform/modules/atlantis/terraform-aws-modules-terraform-aws-atlantis-3eae5e6/main.tf line 401, in module "container_definition_github_gitlab":
 401:   log_options = [
 402:     {
 403:       "awslogs-region"        = "${data.aws_region.current.name}"
 404:       "awslogs-group"         = "${aws_cloudwatch_log_group.atlantis.name}"
 405:       "awslogs-stream-prefix" = "ecs"
 406:     },
 407:   ]

The given value is not suitable for child module variable "log_options"
defined at
.terraform/modules/atlantis.container_definition_github_gitlab/cloudposse-terraform-aws-ecs-container-definition-1e0a6da/variables.tf:91,1-23:
map of any single type required.


Error: Invalid value for module argument

  on .terraform/modules/atlantis/terraform-aws-modules-terraform-aws-atlantis-3eae5e6/main.tf line 433, in module "container_definition_bitbucket":
 433:   log_options = [
 434:     {
 435:       "awslogs-region"        = "${data.aws_region.current.name}"
 436:       "awslogs-group"         = "${aws_cloudwatch_log_group.atlantis.name}"
 437:       "awslogs-stream-prefix" = "ecs"
 438:     },
 439:   ]

The given value is not suitable for child module variable "log_options"
defined at
.terraform/modules/atlantis.container_definition_bitbucket/cloudposse-terraform-aws-ecs-container-definition-1e0a6da/variables.tf:91,1-23:
map of any single type required.
```

**Comment by smiller171** [comment]
Yeah, this module and most of the modules it depends on have not been updated to support Terraform. I've forked them all to work on it and am tackling the last of the bugs I've identified, but it's no small task.

**Comment by antonbabenko** [comment]
Upd: All modules used by this one are already updated to work with Terraform 0.12 and I am working on updating this one right now.

**Comment by antonbabenko** [comment]
[v2.0.0](https://github.com/terraform-aws-modules/terraform-aws-atlantis/releases/v2.0.0) has been just released for Terraform 0.12 users.
